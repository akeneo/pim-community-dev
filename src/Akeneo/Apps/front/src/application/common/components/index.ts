import {Breadcrumb} from './breadcrumb/Breadcrumb';
import {BreadcrumbItem} from './breadcrumb/BreadcrumbItem';
import {ApplyButton} from './button/ApplyButton';
import {GreyButton} from './button/GreyButton';
import {IconButton} from './button/IconButton';
import {ImportantButton} from './button/ImportantButton';
import {DropdownLink, SecondaryActionsDropdownButton} from './button/SecondaryActionsButton';
import {Figure, FigureCaption, FigureImage} from './Figure';
import {Form} from './form/Form';
import {FormGroup} from './form/FormGroup';
import {FormInput} from './form/FormInput';
import {Helper, HelperLink, HelperTitle} from './Helper';
import {InlineHelper} from './InlineHelper';
import {Modal} from './Modal';
import {PageContent} from './PageContent';
import {PageHeader} from './PageHeader';
import {Section} from './Section';
import {Select2} from './Select2';
import {SmallHelper} from './SmallHelper';
import {Select} from './select/Select';

export {
    ApplyButton,
    Breadcrumb,
    BreadcrumbItem,
    DropdownLink,
    Figure,
    FigureCaption,
    FigureImage,
    Form,
    FormGroup,
    FormInput,
    GreyButton,
    Helper,
    HelperLink,
    HelperTitle,
    IconButton,
    ImportantButton,
    InlineHelper,
    Modal,
    PageContent,
    PageHeader,
    SecondaryActionsDropdownButton,
    Section,
    Select2,
    SmallHelper,
    Select,
};
