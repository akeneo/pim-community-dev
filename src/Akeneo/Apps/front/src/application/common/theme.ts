import {ThemedStyledProps} from 'styled-components';

export const theme = {
    color: {
        blue10: '#f5f9fc',
        blue100: '#5992c7',
        grey60: '#f9f9fb',
        grey80: '#d9dde2',
        grey100: '#a1a9b7',
        grey120: '#67768a',
        grey140: '#11324d',
        purple100: '#9452ba',
    },
    fontSize: {
        title: '30px',
        bigger: '17px',
        big: '15px',
        default: '13px',
        small: '11px',
    },
};

export type PropsWithTheme<P = {}> = ThemedStyledProps<P, typeof theme>;
