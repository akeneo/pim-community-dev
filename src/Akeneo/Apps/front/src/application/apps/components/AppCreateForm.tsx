import React, {ChangeEvent, Dispatch, RefObject, useEffect, useReducer, useRef} from 'react';
import {FlowType} from '../../../domain/apps/flow-type.enum';
import {ApplyButton, Form, FormGroup, FormInput} from '../../common';
import {isErr, isOk} from '../../shared/fetch/result';
import {sanitize} from '../../shared/sanitize';
import {Translate} from '../../shared/translate';
import {appWithCredentialsFetched} from '../actions/apps-actions';
import {
    codeGenerated,
    CreateFormAction,
    formIsInvalid,
    formIsValid,
    inputChanged,
    setError,
} from '../actions/create-form-actions';
import {useAppsState} from '../app-state-context';
import {appFormReducer, CreateFormState} from '../reducers/app-form-reducer';
import {CreateAppData, useCreateApp} from '../use-create-app';
import {FlowTypeHelper} from './FlowTypeHelper';
import {FlowTypeSelect} from './FlowTypeSelect';

const initialState: CreateFormState = {
    controls: {
        code: {name: 'code', value: '', errors: {}, dirty: false, valid: false},
        label: {name: 'label', value: '', errors: {}, dirty: false, valid: false},
        flow_type: {
            name: 'flow_type',
            value: FlowType.DATA_SOURCE,
            errors: {},
            dirty: false,
            valid: true,
        },
    },
    valid: false,
};

const useFormValidation = (
    state: CreateFormState,
    dispatch: Dispatch<CreateFormAction>,
    codeInputRef: RefObject<HTMLInputElement>,
    labelInputRef: RefObject<HTMLInputElement>
) => {
    useEffect(() => {
        [codeInputRef, labelInputRef].forEach(inputRef => {
            const input = inputRef.current;
            if (null === input) {
                return;
            }

            const name = input.name;
            if (
                false === input.checkValidity() &&
                0 === Object.keys(state.controls[name].errors).length &&
                true === state.controls[name].dirty
            ) {
                if (input.validity.valueMissing) {
                    dispatch(setError(name, `akeneo_apps.app.constraint.${name}.required`));
                }
                if (input.validity.patternMismatch) {
                    dispatch(setError(name, `akeneo_apps.app.constraint.${name}.invalid`));
                }
            }
        });
    }, [dispatch, codeInputRef, labelInputRef, state.controls]);

    useEffect(() => {
        if (false === state.controls.label.valid || false === state.controls.code.valid) {
            dispatch(formIsInvalid());

            return;
        }
        dispatch(formIsValid());
    }, [dispatch, state.controls.label.valid, state.controls.code.valid]);
};

export const AppCreateForm = () => {
    const [, appsDispatch] = useAppsState();

    const [state, dispatch] = useReducer(appFormReducer, initialState);
    const createNewApp = useCreateApp();

    const codeInputRef = useRef<HTMLInputElement>(null);
    const labelInputRef = useRef<HTMLInputElement>(null);
    useFormValidation(state, dispatch, codeInputRef, labelInputRef);

    useEffect(() => {
        if (true === state.controls.code.dirty) {
            return;
        }

        const value = sanitize(state.controls.label.value);
        if (state.controls.code.value === value) {
            return;
        }

        dispatch(codeGenerated(value));
    }, [dispatch, state.controls.label.value, state.controls.code.value, state.controls.code.dirty]);

    const handleSave = async () => {
        if (false === state.valid) {
            return;
        }
        const data: CreateAppData = {
            code: state.controls.code.value,
            label: state.controls.label.value,
            flow_type: state.controls.flow_type.value as FlowType,
        };

        const result = await createNewApp(data);
        if (isErr(result)) {
            result.error.errors.forEach(({name, reason}) => dispatch(setError(name, reason)));
        }
        if (isOk(result)) {
            appsDispatch(
                appWithCredentialsFetched({
                    code: result.data.code,
                    label: result.data.label,
                    flowType: result.data.flow_type,
                    clientId: result.data.client_id,
                    secret: result.data.secret,
                    username: result.data.username,
                    password: result.data.password,
                })
            );
        }
    };

    const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
        dispatch(inputChanged(event.currentTarget.name, event.currentTarget.value));
    };

    const handleFlowTypeSelect = (flowType: FlowType) => dispatch(inputChanged('flow_type', flowType));

    return (
        <Form>
            <FormGroup controlId='label' label='pim_apps.app.label' errors={Object.keys(state.controls.label.errors)}>
                <FormInput
                    ref={labelInputRef}
                    type='text'
                    name='label'
                    value={state.controls.label.value}
                    onChange={handleChange}
                    required
                    maxLength={100}
                />
            </FormGroup>

            <FormGroup controlId='code' label='pim_apps.app.code' errors={Object.keys(state.controls.code.errors)}>
                <FormInput
                    ref={codeInputRef}
                    type='text'
                    name='code'
                    value={state.controls.code.value}
                    onChange={handleChange}
                    required
                    maxLength={100}
                    pattern='^[0-9a-zA-Z_]+$'
                />
            </FormGroup>

            <FormGroup
                controlId='flow_type'
                label='pim_apps.app.flow_type'
                info={<FlowTypeHelper />}
                required
                errors={Object.keys(state.controls.flow_type.errors)}
            >
                <FlowTypeSelect value={state.controls.flow_type.value as FlowType} onChange={handleFlowTypeSelect} />
            </FormGroup>

            <ApplyButton onClick={handleSave} disabled={false === state.valid}>
                <Translate id='pim_common.save' />
            </ApplyButton>
        </Form>
    );
};
