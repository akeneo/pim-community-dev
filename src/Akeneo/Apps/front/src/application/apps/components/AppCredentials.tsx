import React, {useRef, FC, useContext} from 'react';
import {AppCredentials as AppCredentialsInterface} from '../../../domain/apps/app-credentials.interface';
import {Section, SmallHelper} from '../../common';
import {Translate, TranslateContext} from '../../shared/translate';
import {CredentialList, CredentialListItem} from './CredentialList';
import {RegenerateSecretButton} from './RegenerateSecretButton';
import {IconButton} from '../../common';
import {DuplicateIcon} from '../../common/icons';
import {useNotify, NotificationLevel} from '../../shared/notify';
import styled from 'styled-components';

interface Props {
    code: string;
    appCredentials: AppCredentialsInterface;
}

export const AppCredentials: FC<Props> = ({code, appCredentials}: Props) => {
    const notify = useNotify();
    const translate = useContext(TranslateContext);

    const clientIdRef = useRef<HTMLElement>(null);
    const secretRef = useRef<HTMLElement>(null);

    const handleCopy = (element: HTMLElement) => {
        const selection = window.getSelection();
        if (null === selection) {
            return;
        }

        const range = document.createRange();
        range.selectNodeContents(element);

        selection.removeAllRanges();
        selection.addRange(range);

        document.execCommand('copy');

        selection.removeAllRanges();
    };

    const handleClientIdCopy = () => {
        if (null === clientIdRef.current) {
            return;
        }
        handleCopy(clientIdRef.current);

        notify(NotificationLevel.INFO, translate('akeneo_apps.edit_app.credentials.flash.copied', {name: 'Client ID'}));
    };

    const handleSecretCopy = () => {
        if (null === secretRef.current) {
            return;
        }
        handleCopy(secretRef.current);

        notify(NotificationLevel.INFO, translate('akeneo_apps.edit_app.credentials.flash.copied', {name: 'Secret'}));
    };

    return (
        <>
            <Section title={<Translate id='akeneo_apps.edit_app.credentials.title' />} />
            <div>
                <SmallHelper>
                    <Translate id='akeneo_apps.edit_app.credentials.helper' />
                </SmallHelper>
            </div>

            <CredentialList>
                <CredentialListItem
                    label={<Translate id='akeneo_apps.app.client_id' />}
                    action={
                        <IconButton
                            onClick={handleClientIdCopy}
                            title={translate('akeneo_apps.edit_app.credentials.action.copy')}
                        >
                            <DuplicateIcon />
                        </IconButton>
                    }
                >
                    <span ref={clientIdRef}>{appCredentials.clientId}</span>
                </CredentialListItem>
                <CredentialListItem
                    label={<Translate id='akeneo_apps.app.secret' />}
                    action={
                        <ActionList>
                            <IconButton
                                onClick={handleSecretCopy}
                                title={translate('akeneo_apps.edit_app.credentials.action.copy')}
                            >
                                <DuplicateIcon />
                            </IconButton>
                            <RegenerateSecretButton code={code} />
                        </ActionList>
                    }
                >
                    <span ref={secretRef}>{appCredentials.secret}</span>
                </CredentialListItem>
            </CredentialList>
        </>
    );
};

const ActionList = styled.div`
    > * {
        margin: 0 5px;

        :first-child {
            margin-left: 0;
        }
        :last-child {
            margin-right: 0;
        }
    }
`;
