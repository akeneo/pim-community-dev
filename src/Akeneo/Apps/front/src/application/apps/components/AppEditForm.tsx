import {FormikProps} from 'formik';
import React from 'react';
import {App} from '../../../domain/apps/app.interface';
import {FormGroup, FormInput, Section} from '../../common';
import {Translate} from '../../shared/translate';
import {FormValues} from '../pages/AppEdit';
import {FlowTypeHelper} from './FlowTypeHelper';
import {FlowTypeSelect} from './FlowTypeSelect';
import styled from 'styled-components';

interface Props {
    app: App;
    formik: FormikProps<FormValues>;
}

export const AppEditForm = ({app, formik: {values, handleChange, setFieldValue, errors}}: Props) => {
    return (
        <>
            <Section title={<Translate id='pim_apps.edit_app.subtitle' />} />

            <br />

            <Container>
                <FormGroup controlId='code' label='pim_apps.app.code'>
                    <FormInput type='text' defaultValue={app.code} disabled />
                </FormGroup>

                <FormGroup
                    controlId='label'
                    label='pim_apps.app.label'
                    errors={errors.label ? [errors.label] : undefined}
                >
                    <FormInput
                        type='text'
                        name='label'
                        value={values.label}
                        onChange={handleChange}
                        required
                        maxLength={100}
                    />
                </FormGroup>

                <FormGroup controlId='flow_type' label='pim_apps.app.flow_type' info={<FlowTypeHelper />}>
                    <FlowTypeSelect
                        value={values.flowType}
                        onChange={flowType => setFieldValue('flowType', flowType)}
                    />
                </FormGroup>
            </Container>
        </>
    );
};

const Container = styled.div`
    width: 100%;
    max-width: 400px;
`;
