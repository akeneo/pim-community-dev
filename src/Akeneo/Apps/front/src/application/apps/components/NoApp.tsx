import React from 'react';
import styled from 'styled-components';
import imgUrl from '../../common/assets/illustrations/api.svg';
import {PropsWithTheme} from '../../common/theme';
import {Translate} from '../../shared/translate';

const Container = styled.div`
    text-align: center;
`;

const Title = styled.div`
    height: 36px;
    font-size: ${({theme}: PropsWithTheme) => theme.fontSize.title};
    color: ${({theme}: PropsWithTheme) => theme.color.grey140};
`;

const Message = styled.div`
    height: 21px;
    font-size: ${({theme}: PropsWithTheme) => theme.fontSize.bigger};
`;

const Link = styled.a`
    color: #9452ba;
    cursor: pointer;
    text-decoration: underline ${({theme}: PropsWithTheme) => theme.color.purple100};
`;

const Image = styled.img`
    width: 256px;
`;

export const NoApp = ({onCreate}: {onCreate: () => void}) => (
    <Container>
        <Image src={imgUrl} />
        <Title>
            <Translate id='pim_apps.no_app.title' />
        </Title>
        <Message>
            <Translate id='pim_apps.no_app.message' />
            &nbsp;
            <Link onClick={onCreate}>
                <Translate id='pim_apps.no_app.message_link' />
            </Link>
        </Message>
    </Container>
);
