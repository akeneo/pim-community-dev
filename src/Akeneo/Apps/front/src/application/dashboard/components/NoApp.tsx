import React from 'react';
import styled from 'styled-components';
import imgUrl from '../../common/assets/illustrations/api.svg';
import {PropsWithTheme} from '../../common/theme';
import {Translate} from '../../shared/translate';

export const NoApp = () => (
    <Container>
        <Image src={imgUrl} />
        <Title>
            <Translate id='pim_apps.no_app.title' />
        </Title>
    </Container>
);

const Container = styled.div`
    text-align: center;
`;

const Image = styled.img`
    width: 256px;
`;

const Title = styled.div`
    height: 36px;
    font-size: ${({theme}: PropsWithTheme) => theme.fontSize.title};
    color: ${({theme}: PropsWithTheme) => theme.color.grey140};
`;
