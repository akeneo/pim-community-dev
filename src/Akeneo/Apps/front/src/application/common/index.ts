import {Breadcrumb} from './components/breadcrumb/Breadcrumb';
import {BreadcrumbItem} from './components/breadcrumb/BreadcrumbItem';
import {ApplyButton} from './components/button/ApplyButton';
import {Figure, FigureCaption, FigureImage} from './components/Figure';
import {Form} from './components/form/Form';
import {FormGroup} from './components/form/FormGroup';
import {FormInput} from './components/form/FormInput';
import {Helper, HelperLink, HelperTitle} from './components/Helper';
import {Loading} from './components/Loading';
import {Modal} from './components/Modal';
import {Page} from './components/Page';
import {PageHeader} from './components/PageHeader';
import {Section} from './components/Section';
import {Select2} from './components/Select2';

export {
    ApplyButton,
    Breadcrumb,
    BreadcrumbItem,
    Figure,
    FigureCaption,
    FigureImage,
    Form,
    FormGroup,
    FormInput,
    Helper,
    HelperLink,
    HelperTitle,
    Loading,
    Modal,
    Page,
    PageHeader,
    Section,
    Select2,
};
