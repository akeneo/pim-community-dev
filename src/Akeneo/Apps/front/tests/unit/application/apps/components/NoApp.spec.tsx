import {mount} from 'enzyme';
import * as React from 'react';
import {create} from 'react-test-renderer';
import {ThemeProvider} from 'styled-components';
import {NoApp} from '../../../../../src/application/apps/components/NoApp';
import {theme} from '../../../../../src/application/common/theme';

describe('NoApp', () => {
    it('should render', () => {
        const component = create(
            <ThemeProvider theme={theme}>
                <NoApp onCreate={() => undefined} />
            </ThemeProvider>
        );

        expect(component.toJSON()).toMatchSnapshot();
    });

    it('should call `onCreate` when the create app shortcut is clicked', () => {
        const handleCreate = jest.fn();

        const component = mount(
            <ThemeProvider theme={theme}>
                <NoApp onCreate={handleCreate} />
            </ThemeProvider>
        );

        component.find('a').simulate('click');
        expect(handleCreate).toBeCalled();
    });
});
