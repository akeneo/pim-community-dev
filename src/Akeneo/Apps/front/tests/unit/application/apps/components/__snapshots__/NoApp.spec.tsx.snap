// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoApp should render 1`] = `
<div
  className="sc-bdVaJa kmXoBq"
>
  <img
    className="sc-ifAKCX isanNQ"
    src="test-file-stub"
  />
  <div
    className="sc-bwzfXH fSQTvU"
  >
    pim_apps.no_app.title
  </div>
  <div
    className="sc-htpNat hTaSGc"
  >
    pim_apps.no_app.message
    Â 
    <a
      className="sc-bxivhb goiCAm"
      onClick={[Function]}
    >
      pim_apps.no_app.message_link
    </a>
  </div>
</div>
`;
